File: python/src/azoth/eos/reference/hydrate_fraction.py
"""``eos.hydrate_fraction``: the Python reference.

The mirror of ``crates/azoth-eos/src/hydrate_fraction.rs``:

```text
f_i = z_i - beta w_i                   the fluid the hydrate leaves, as mole numbers
solve  beta = min_i z_i / w_i          the fraction at which a component is exhausted
```

NeqSim's ``TPHydrateFlash`` computes the same quantity and **its state is not one**: the
hydrate it leaves has mole fractions summing to ``1.1201``, and the phases hold ``+0.0874``
water and ``-0.0734`` methane against the feed. The divergence is recorded upstream and in
the spec; this is the solve that closes.

What decides between no hydrate and all of the water is the same objective
:mod:`azoth.eos.reference.hydrate_formation_temperature` drives to zero, read at the **feed**:
the objective is monotone in the fraction, so a feed above its formation temperature has no
hydrate at all, and one below it takes the whole of the water there is.
"""

from __future__ import annotations

import math

from azoth.core.errors import InvalidInputError, SolverNotConvergedError
from azoth.core.range import apply_checks, checks_for
from azoth.core.result import HydrateFractionResult, HydrateStructure, PtFlashResult
from azoth.core.units import Q, from_si, input_to_si
from azoth.core.warnings import Warning
from azoth.eos.mixture import Mixture
from azoth.eos.reference import _hydrate
from azoth.eos.reference._mixture_state import phase_state, reduced_parameters
from azoth.eos.reference.pt_flash import pt_flash

MODEL_ID = "eos.hydrate_fraction"


def _reference_water_fugacity(eos: str, t: float, p: float) -> float:
    """The fugacity of pure water at a state on the cubic the fluid runs, in Pa."""
    from azoth.eos import components as databank

    water, _ = databank.mixture_of(["water"], eos=eos)
    reduced = reduced_parameters(water, t, p)
    state = phase_state(reduced, water.kij, [1.0], liquid=True)
    return math.exp(state.ln_phi[0]) * p


def _guest_fugacities(flash: PtFlashResult, fluid: list[float], p: float) -> list[float]:
    """The guests' fugacities at a flash, in Pa, from the phase ``setFug`` reads."""
    composition = list(flash.y) if flash.phase == "two_phase" else list(fluid)
    coefficients = flash.ln_phi_liquid if flash.phase == "all_liquid" else flash.ln_phi_vapour
    return [
        fraction * math.exp(coefficient) * p
        for fraction, coefficient in zip(composition, coefficients, strict=True)
    ]


def _fluid_at(z: list[float], in_hydrate: list[float], beta: float) -> list[float]:
    """The fluid the hydrate leaves at one fraction, as mole fractions.

    ``f_i = z_i - beta w_i`` is that fluid's own mole numbers, summing to ``1 - beta``. A
    component the hydrate has taken all of comes out at zero rather than negative, which is
    the state the bound is *defined* by.
    """
    fluid = [max(total - beta * taken, 0.0) for total, taken in zip(z, in_hydrate, strict=True)]
    total = sum(fluid)
    return [fraction / total for fraction in fluid]


def _maximum_fraction(z: list[float], in_hydrate: list[float]) -> float:
    """The largest fraction a feed's own composition allows: ``min_i z_i / w_i``."""
    ratios = [total / taken for total, taken in zip(z, in_hydrate, strict=True) if taken > 1.0e-12]
    return min(ratios) if ratios else math.inf


def hydrate_fraction(
    components: list[str], T: Q, P: Q, z: list[float], eos: str = "srk"
) -> HydrateFractionResult:
    """The fraction of a feed that is hydrate at a temperature and pressure.

    **The components cross by name**, because the hydrate's guest tables are keyed by name
    and a :class:`~azoth.eos.mixture.Component` carries none.

    Args:
        components: the substances, by name. At least one must be water and one a former.
        T: absolute temperature; it must be below the formation temperature, or the answer
            is zero.
        P: absolute pressure.
        z: the overall mole fractions. Checked rather than renormalised.
        eos: the cubic the fluid runs, which is also the one the hydrate's reference water
            phase is built from.

    Raises:
        InvalidInputError: if the fluid has no water, or nothing in it occupies a cage.
        OutOfRangeError: if ``T``, ``P`` or a trial's cavity sum has no value.
        SolverNotConvergedError: if the fraction and the composition do not settle.
    """
    from azoth import _models_gen
    from azoth.eos import components as databank

    spec = _models_gen.model(MODEL_ID)
    checks = checks_for(spec)
    warnings: list[Warning] = []
    t_si = input_to_si(spec, "T", T)
    p_si = input_to_si(spec, "P", P)
    apply_checks(checks.on_input, {"T": t_si, "P": p_si}.get, warnings)

    hydration = _hydrate.hydration_for(components)
    if hydration.water_index is None:
        raise InvalidInputError("components", "the hydration carries no water index")
    water_index = hydration.water_index
    mixture, _ = databank.mixture_of(components, eos=eos)

    algorithm = spec["algorithm"]
    tolerance = float(algorithm["tolerance"])
    max_iterations = int(algorithm["max_iterations"])

    def cages(fluid: list[float]) -> tuple[list[float], int, float, list[float]]:
        """One trial: flash the fluid, fill the cages by what the flash leaves."""
        flash = pt_flash(mixture, from_si(t_si, "K"), from_si(p_si, "Pa"), fluid)
        fugacities = _guest_fugacities(flash, fluid, p_si)
        reference = _reference_water_fugacity(eos, t_si, p_si)
        structure, coefficient = _hydrate.stable_structure(
            hydration.guests, fugacities, t_si, p_si, reference
        )
        return (
            _hydrate.composition(hydration.guests, fugacities, structure, t_si, water_index),
            structure,
            coefficient,
            fugacities,
        )

    # **The feed's own state decides whether there is any hydrate at all.** The objective is
    # only read here: at the bound the fluid holds no water, so the ratio has no value at it,
    # and the fixed point below needs the cages rather than the residual.
    iterations = 1
    composition, structure, coefficient, fugacities = cages(list(z))
    residual = math.log(coefficient * p_si / fugacities[water_index])
    if residual > 0.0:
        return _finish(
            mixture, t_si, p_si, z, composition, structure, 0.0, residual, iterations, warnings
        )

    # **Below it the hydrate takes the water there is**, and the fraction and the cages are
    # one fixed point: the fraction is ``min_i z_i/w_i`` from the cages' own composition, and
    # the cages are filled from the fluid that fraction leaves.
    beta = _maximum_fraction(z, composition)
    for step in range(2, max_iterations + 1):
        if not math.isfinite(beta):
            break
        fluid = _fluid_at(z, composition, beta)
        composition, structure, _coefficient, _fugacities = cages(fluid)
        iterations = step
        following = _maximum_fraction(z, composition)
        if abs(following - beta) < tolerance:
            return _finish(
                mixture,
                t_si,
                p_si,
                z,
                composition,
                structure,
                following,
                residual,
                iterations,
                warnings,
            )
        beta = following

    raise SolverNotConvergedError(iterations=iterations, residual=beta, tolerance=tolerance)


def _finish(
    mixture: Mixture,
    t_si: float,
    p_si: float,
    z: list[float],
    composition: list[float],
    structure: int,
    beta: float,
    residual: float,
    iterations: int,
    warnings: list[Warning],
) -> HydrateFractionResult:
    """The answer, with the material balance it exists to keep measured and reported."""
    fluid = _fluid_at(z, composition, beta)
    flash = pt_flash(mixture, from_si(t_si, "K"), from_si(p_si, "Pa"), fluid)
    split = flash.beta if flash.beta is not None else 0.0
    if flash.phase == "two_phase":
        x, y = flash.x, flash.y
    else:
        x = y = tuple(fluid)
    balance_error = 0.0
    for i, total in enumerate(z):
        left = ((1.0 - split) * x[i] + split * y[i]) * (1.0 - beta)
        balance_error = max(balance_error, abs(beta * composition[i] + left - total))

    return HydrateFractionResult(
        beta=beta,
        structure=(
            HydrateStructure.STRUCTURE_I if structure == 0 else HydrateStructure.STRUCTURE_II
        ),
        balance_error=balance_error,
        iterations=iterations + 1,
        residual=residual,
        warnings=tuple(warnings),
    )


__all__ = ["hydrate_fraction"]

File: crates/azoth-eos/src/hydrate_fraction.rs
//! `eos.hydrate_fraction` - how much of a feed is hydrate at a state.
//!
//! ```text
//! f_i = z_i - beta w_i                   the fluid the hydrate leaves, as mole numbers
//! solve  beta = min_i z_i / w_i          the fraction at which a component is exhausted
//! ```
//!
//! NeqSim's `TPHydrateFlash` computes the same quantity from the same construction, and the
//! two agree: measured at 288.15 K and 100 bara its balance error is `7.3e-13` on methane and
//! `0.0` on water, and its fraction moves with temperature as this one's does. **The solve is
//! what differs** - it optimises over the water extent where this takes the bound directly.
//!
//! # The objective is flat, and the answer is where the water runs out
//!
//! Measured in the fraction, at 288.15 K and 100 bara on the probe's feed: the aqueous phase
//! pins water's fugacity, so `ln(f_w^hydrate/f_w^fluid)` sits at `-0.071993842` at a fraction
//! of zero and has moved only to `-0.069796728` by the time the fluid holds a trace - a
//! change of `2e-3` across the whole range, with the fluid's own water fugacity identical to
//! nine digits. Past that trace the aqueous phase is gone and the objective climbs to `+inf`
//! at the state where the last water molecule has left. **So it never crosses zero: the
//! hydrate is stable at every fraction up to the one where the water is exhausted, and that
//! fraction is the answer.**
//!
//! This is why the two earlier readings of the probe looked like physics and were not. Its
//! objective appears flat in the fraction - it is, and the flatness is real; what was an
//! artefact is where NeqSim's flat part *ends*, because its composition is wrong. And its
//! fraction appears temperature-independent, which it is, because its bound is the constant
//! `z_water / (46/54)`.
//!
//! # What decides zero
//!
//! The same objective measured at the **feed**, where the hydrate has taken nothing. A feed
//! above its formation temperature has no hydrate at all, and that is exactly the condition
//! [`crate::hydrate_formation_temperature`] solves for - so the two models are one statement
//! read in two directions, and the residual here is the quantity the other drives to zero.
//!
//! # The composition is the cages' own
//!
//! From both cavity types of the stable structure ([`crate::hydrate::composition`]).
//! The balance closes by construction: the fluid is `z_i - beta w_i`, which sums to `1 - beta` for any `w` that
//! sums to one, so no phase can hold matter the feed did not have.
//!
//! # The two loops
//!
//! The composition and the fluid are mutually dependent: the cages fill by the fluid's
//! fugacities, and the fluid is what the cages have already taken from. The fraction and the
//! composition are one fixed point - `beta = min_i z_i/w_i` is a function of `w`, and the
//! fluid it leaves is a function of both - so the iteration is over that pair, and each step
//! flashes a state the hydrate has already taken the water out of.

use azoth_core::units::{Pressure, ThermodynamicTemperature, kelvins};
use azoth_core::{AzothError, Result, apply_checks};

use crate::Cubic;
use crate::databank;
use crate::hydrate::{self, Hydration};
use crate::mixture::{Mixture, RootSide};
use crate::model_gen;
use crate::pt_flash::pt_flash;
use crate::results::{HydrateFractionResult, HydrateStructure, Phase, PtFlashResult};

/// The fugacity of pure water at a state on the cubic the fluid runs, in Pa.
///
/// # Errors
/// * [`AzothError`] from the water-only mixture's own state.
fn reference_water_fugacity(eos: Cubic, t: f64, p: f64) -> Result<f64> {
    let (water, _) = databank::mixture_of(&["water"], eos, None)?;
    let reduced = water.reduced_parameters(kelvins(t), azoth_core::units::pascals(p))?;
    let state = water.phase_state(&reduced, &[1.0], RootSide::Liquid)?;
    Ok(state.ln_phi[0].exp() * p)
}

/// The guests' fugacities at a flash, in Pa, from the phase NeqSim's `setFug` reads.
fn guest_fugacities(flash: &PtFlashResult, fluid: &[f64], p: f64) -> Vec<f64> {
    let composition = if flash.phase == Phase::TwoPhase {
        flash.y.clone()
    } else {
        fluid.to_vec()
    };
    let ln_phi = if flash.phase == Phase::AllLiquid {
        &flash.ln_phi_liquid
    } else {
        &flash.ln_phi_vapour
    };
    composition
        .iter()
        .zip(ln_phi)
        .map(|(fraction, coefficient)| fraction * coefficient.exp() * p)
        .collect()
}

/// The hydrate the fluid at one state builds, and the water fugacity that comes with it.
struct Trial {
    /// The hydrate's mole fractions, from the cages the fluid fills.
    composition: Vec<f64>,
    /// The stable structure, as the index [`hydrate`] uses.
    structure: usize,
    /// The guests' fugacities the cages were filled from, in Pa.
    fugacities: Vec<f64>,
    /// The hydrate's water fugacity divided by the pressure.
    coefficient: f64,
}

/// One trial: flash the fluid, fill the cages by what the flash leaves, keep the structure.
fn trial(
    mixture: &Mixture,
    t: ThermodynamicTemperature,
    p: Pressure,
    fluid: &[f64],
    hydration: &Hydration,
) -> Result<Trial> {
    let water_index = hydration
        .water_index
        .expect("the resolver refuses no water");
    let flash = pt_flash(mixture, t, p, fluid)?;
    let fugacities = guest_fugacities(&flash, fluid, p.value);
    let reference = reference_water_fugacity(mixture.cubic(), t.value, p.value)?;
    let (structure, coefficient) = hydrate::stable_structure(
        &hydration.guests,
        &fugacities,
        hydration.model,
        t.value,
        p.value,
        reference,
    )?;
    Ok(Trial {
        composition: hydrate::composition(
            &hydration.guests,
            &fugacities,
            hydration.model,
            structure,
            t.value,
            water_index,
        ),
        structure,
        fugacities,
        coefficient,
    })
}

/// `ln(f_w^hydrate / f_w^fluid)` at a trial, in Pa over Pa.
fn objective(trial: &Trial, p: f64, water_index: usize) -> f64 {
    (trial.coefficient * p / trial.fugacities[water_index]).ln()
}

/// The largest fraction a feed's own composition allows: `min_i z_i / w_i`.
///
/// A fraction past it asks some component for more than the feed holds, so this is where a
/// component is exhausted - and for a water-limited feed the water is what is.
fn maximum_fraction(z: &[f64], in_hydrate: &[f64]) -> f64 {
    z.iter()
        .zip(in_hydrate)
        .filter(|(_, taken)| **taken > 1.0e-12)
        .map(|(total, taken)| total / taken)
        .fold(f64::INFINITY, f64::min)
}

/// The fluid the hydrate leaves at one fraction, as mole fractions.
///
/// `f_i = z_i - beta w_i` is that fluid's own mole numbers, summing to `1 - beta`. A
/// component the hydrate has taken all of comes out at zero rather than negative, which is
/// the state the bound is *defined* by: it is what makes the water-free trial a state the
/// hydrate has really left rather than one it has overrun.
fn fluid_at(z: &[f64], in_hydrate: &[f64], beta: f64) -> Vec<f64> {
    let mut fluid: Vec<f64> = z
        .iter()
        .zip(in_hydrate)
        .map(|(total, taken)| (total - beta * taken).max(0.0))
        .collect();
    let sum: f64 = fluid.iter().sum();
    for fraction in &mut fluid {
        *fraction /= sum;
    }
    fluid
}

/// The fraction of a feed that is hydrate at a temperature and pressure.
///
/// # Errors
/// * [`AzothError::InvalidInput`] if the mixture was not resolved for a hydrate.
/// * [`AzothError::OutOfRange`] if a trial's cavity sum has no value.
/// * [`AzothError::SolverNotConverged`] if the fraction and the composition do not settle.
pub fn hydrate_fraction(
    mixture: &Mixture,
    t: ThermodynamicTemperature,
    p: Pressure,
    z: &[f64],
) -> Result<HydrateFractionResult> {
    let spec = &model_gen::HYDRATE_FRACTION_SPEC;
    let mut warnings = Vec::new();
    apply_checks(
        spec.input_checks(),
        |quantity| match quantity {
            "T" => Some(t.value),
            "P" => Some(p.value),
            _ => None,
        },
        &mut warnings,
    )?;
    let algorithm = crate::algorithm_of(spec)?;

    let hydration = mixture.hydration().ok_or_else(|| {
        AzothError::invalid_input(
            "components",
            "this mixture was not resolved for a hydrate; `hydrate_mixture_of` is what \
             attaches the guests' tables"
                .to_string(),
        )
    })?;
    let water_index = hydration
        .water_index
        .expect("the resolver refuses no water");

    // **The feed's own state decides whether there is any hydrate at all.** The objective is
    // monotonically increasing in the fraction - the fluid's water fugacity falls as the
    // water leaves it - so a feed whose objective is already positive has no root anywhere
    // and no hydrate, which is the same statement as its being above its formation
    // temperature. Measured, the two models cross zero together: the residual here is
    // `-0.003202` at 293 K and `+0.010844` at 294 K, and the formation temperature is
    // `293.2277` K.
    let mut iterations = 1;
    let feed = trial(mixture, t, p, z, hydration)?;
    let residual = objective(&feed, p.value, water_index);
    if residual > 0.0 {
        return finish(mixture, t, p, z, &feed, 0.0, residual, iterations, warnings);
    }

    // **Below it the hydrate takes the water there is**, and the fraction and the cages are
    // one fixed point: the fraction is `min_i z_i/w_i` from the cages' own composition, and
    // the cages are filled from the fluid that fraction leaves. The start is the feed's own
    // fluid - a hydrate in equilibrium with nothing - and each step flashes a state with the
    // water already taken out, which is the state at the answer rather than one near it.
    let mut hydrated = feed;
    let mut beta = maximum_fraction(z, &hydrated.composition);
    for step in 2..=algorithm.max_iterations {
        if !beta.is_finite() {
            break;
        }
        let fluid = fluid_at(z, &hydrated.composition, beta);
        hydrated = trial(mixture, t, p, &fluid, hydration)?;
        iterations = step;
        let next = maximum_fraction(z, &hydrated.composition);
        if (next - beta).abs() < algorithm.tolerance {
            return finish(
                mixture, t, p, z, &hydrated, next, residual, iterations, warnings,
            );
        }
        beta = next;
    }

    Err(AzothError::SolverNotConverged {
        iterations,
        residual: beta,
        tolerance: algorithm.tolerance,
    })
}

/// The answer, with the material balance it exists to keep measured and reported.
///
/// The fluid is `z_i - beta w_i` and the hydrate's fraction is `beta w_i`, so the
/// recombination is exact by construction whatever the composition is. What is *measured* is
/// the state the flash leaves: `sum_p beta_p x_ip` over the fluid's own phases and the
/// hydrate, against the feed's `z_i`. That is the same quantity NeqSim's state misses on
/// water by `0.0874`, and it tests this library's flash as much as this model.
#[allow(clippy::too_many_arguments)]
fn finish(
    mixture: &Mixture,
    t: ThermodynamicTemperature,
    p: Pressure,
    z: &[f64],
    hydrated: &Trial,
    beta: f64,
    residual: f64,
    iterations: u32,
    warnings: Vec<azoth_core::Warning>,
) -> Result<HydrateFractionResult> {
    let fluid = fluid_at(z, &hydrated.composition, beta);
    let flash = pt_flash(mixture, t, p, &fluid)?;
    // `beta` is the **vapour** fraction, so the fluid's own composition is
    // `(1 - split) x + split y`. At the answer the fluid is the water-free one the hydrate
    // leaves, which flashes single phase, so this split is the general form rather than the
    // one the bound is read at.
    let split = flash.beta.unwrap_or(0.0);
    let (x, y) = match flash.phase {
        Phase::TwoPhase => (flash.x.clone(), flash.y.clone()),
        _ => (fluid.clone(), fluid.clone()),
    };
    let mut balance_error: f64 = 0.0;
    for i in 0..z.len() {
        let left = ((1.0 - split) * x[i] + split * y[i]) * (1.0 - beta);
        let total = beta * hydrated.composition[i] + left;
        balance_error = balance_error.max((total - z[i]).abs());
    }

    Ok(HydrateFractionResult {
        beta,
        structure: if hydrated.structure == 1 {
            HydrateStructure::StructureIi
        } else {
            HydrateStructure::StructureI
        },
        balance_error,
        iterations: iterations + 1,
        residual,
        warnings,
    })
}
